### ISSUE TITLE
Race condition for relational reordering

### ISSUE THREAD

**Issue by Cpaczek** [strapi/strapi#15558]
### Required System information

Node.js version: v18
Strapi version: 4.5.6
Database: All (testing on maria cluster)
Operating system: Windows 10

### Describe the bug
https://github.com/strapi/strapi/blob/127c04705751139bb32805e79b6f8619d92b6d34/packages/core/database/lib/entity-manager/index.js#L605-L634

This uses a select to get the most recent order value and adds 1 to it. This could cause a race condition if you try and add a permission order row twice at the same time because it could depend on old max values.

This is related to https://github.com/strapi/strapi/issues/15539 making this new issue as it turned into a performance discussion instead of race condition.

### Steps to reproduce the behavior

```js
let arr = []
for (let i = 0; i < 100; i++) {
  arr.push(i.toString())
}
Promise.all(arr.map((i) => strapi.entityService.create("api::blog.blog", {
  data: {
    tester: i,
    comments: [1]
  }
})))
```
Because we are doing promise.all all the creates will happen at the same time triggering the race condition. 

This causes order numbers to be grouped when they should be sequential:
![image](https://user-images.githubusercontent.com/25715982/214429762-1ae7704e-a2f0-4649-8706-53853af17382.png)

The main issue is that it causes some weird logic when displaying the relations where all of them don't show up:
https://i.imgur.com/hZXaOqP.gif

**Comment by Marc-Roig** [comment]
To sumarize what was discussed about this in the previous thread https://github.com/strapi/strapi/issues/15539 👍 

- Race condition occurs because the application calculates the order, so creating multiple relationships in parallel can result in repeated order values.

Potential solutions: 
- Compound auto-increment value created by the DB
- Implement locking here to prevent race conditions

**Comment by Marc-Roig** [comment]
Testing and looking at https://github.com/strapi/strapi/pull/15554 I think this issue might be fixed too.

Every insert reorders all the order values of the entity relationship. 

Testing with your setup seems that I can not reproduce it anymore. We will release it inside 4.6.0 today.

**Comment by Marc-Roig** [comment]
Testing and looking at https://github.com/strapi/strapi/pull/15554 I think this issue might be mitigated a little bit.

Every insert reorders all the order values of the entity relationship. 

Testing with your setup it seems I do see repeated order columns at first, but when I edit one of the relations in the content manager all the order values are regenerated. We will release it inside 4.6.0 today.

**Comment by Cpaczek** [comment]
> Testing and looking at #15554 I think this issue might be mitigated a little bit.

It still is something that should probably be fixed (and needs to be fixed for CRDB support) as with Cockroach DB it won't allow for duplicate values to be inserted (I assume because of their concurrency control).

**Comment by Marc-Roig** [comment]
We have been discussing about this issues and race conditions in strapi.

**There is no straightforward method of locking a table using knex js**. Doing so it would require using raw queries and evaluating how each database handles locking (sqlite, postgres, mysql, cockrach all of them behaves differently).

**We think this complexity would fix some of the race conditions but not all of them**. Specially things related with relations.

Our take currently on this issue is updates should not be made in parallel like in this issue description. Each entity should be awaited one by one, and we should be encouraging that in our documentation.

So:
- If possible, could you check @Cpaczek what happens with cockroach db?
- Our take on this is, using the ContentAPI with sequential updates , this should never happen (very low chance). And even if it happens, for relations, the order value will be regenerated as soon as you make another update. So we think it's not worth the complexity it requires.

wdyt @Cpaczek ?

**Comment by Cpaczek** [comment]
> We have been discussing about this issues and race conditions in strapi.
> 
> **There is no straightforward method of locking a table using knex js**. Doing so it would require using raw queries and evaluating how each database handles locking (sqlite, postgres, mysql, cockrach all of them behaves differently).
> 
> **We think this complexity would fix some of the race conditions but not all of them**. Specially things related with relations.
> 
> Our take currently on this issue is updates should not be made in parallel like in this issue description. Each entity should be awaited one by one, and we should be encouraging that in our documentation.
> 
> So:
> 
> * If possible, could you check @Cpaczek what happens with cockroach db?
> * Our take on this is, using the ContentAPI with sequential updates , this should never happen (very low chance). And even if it happens, for relations, the order value will be regenerated as soon as you make another update. So we think it's not worth the complexity it requires.
> 
> wdyt @Cpaczek ?

@Marc-Roig The issue is that if you hit a race condition (i.e from doing a Promise.all()) it won't allow relations to be added to the links table as CRDB enforces the uniqueness of the foreign keys in the links table. 

From my understanding, the race condition will only be hit if you are directly interacting with the entity service, not from the content-api. So as long as you don't do promise.all on queries then it should all be good.

The proper solution is not table locking though, it's compound auto-increment but I don't think that has super wide support. The race condition is that you are doing a query and then incrementing it with js instead of letting the database increment it.

**Comment by Marc-Roig** [comment]
@cpaczek Agree, for this issue, compound auto-increment could work. Still there are other places where race conditions could occur.
Example: https://github.com/strapi/strapi/issues/13155

Relations are tricky to handle and we might be missing even more cases of race conditions (updating compos/dz in parallel for instance)

**Comment by cpaczek** [comment]
> Still there are other places where race conditions could occur.
> Example: #13155

@Marc-Roig Compound auto increment should eliminate all race conditions related to relations right? Or are you talking about somewhere else?

Currently, if you are only running 1 Strapi instance the race conditions only arise if you are doing something like promise.all where you are running the relation increment code concurrently. This means that good code can prevent these race conditions. However, if you are horizontally scaling strapi i.e more than 1 instance you can trigger race conditions even from the content API.

**Comment by Marc-Roig** [comment]
> > Still there are other places where race conditions could occur.
> 
> > Example: #13155
> 
> 
> 
> @Marc-Roig Compound auto increment should eliminate all race conditions related to relations right? Or are you talking about somewhere else?
> 
> 
> 
> Currently, if you are only running 1 Strapi instance the race conditions only arise if you are doing something like promise.all where you are running the relation increment code concurrently. This means that good code can prevent these race conditions. However, if you are horizontally scaling strapi i.e more than 1 instance you can trigger race conditions even from the content API. 
> 
> 
> 
> 

Maybe I lack knowledge on compound auto increment, but how would it solve having duplicate one to one relations? We apply the constraint of only having one of those per pair of entities in our application.
